### PR TITLE
Add a crew monitoring server

### PR DESCRIPTION
**Pull request by juliangiebel**
## About the PR 
* Adds crew monitoring servers
  * Only one is active at a time. If the one that is currently active loses power or gets removed / destroyed another available server will take over
* Changes suit sensors to not broadcast anymore
* Adds a random offset to each suit sensors update time

Suit sensors don't just broadcast their status anymore! 
Instead they retrieve the currently active crew monitoring servers address from the crew monitoring system and send their status directly to that server.

That server is then periodically broadcasting the status of all "connected" suit sensors in one packet on a different frequency.

So we went from a huge amount of broadcasting devices to just one.
  
**Changelog**

:cl:
- add: The crew monitoring server. It relays the status of suit sensors to crew monitors.

### CODE REVIEW

**Comment by moonheart08** [comment]
Does this work when multiple stations are loaded.

**Comment by juliangiebel** [comment]
It doesn't break but you still see all crew members regardles of what station they are on.

I could restrict all of the monitoring equipment to only send and receive only for the grid they are on and have one active server per grid.
I would only need to figure out how to make sure suit sensors work of station too.
A frequency offset might work but that will break with to many stations.

**Comment by Fishfish458** [comment]
> 

I assume the ultimate goal would be distance based from server but i don't think that is cheap to process currently.

**Comment by Emisse** [comment]
would there be a way to tie crew to the station grid they join in on

**Comment by Macoron** [comment]
I understand this solution for performance reasons, but have some feedback gameplay-wise:

- When you leave 500 tiles range from a server, your suit sensor won't work. And even if you build another one on your grid it won't help, because all crew monitorings are connected to one world server. So if you have two stations with 500 tiles distance between them - one station won't be able to use their crew monitoring devices. 
- There is no way to build a server. You probably need construction graph and circuit.
- Mappers should map it asap after this PR merge.
- Crew monitoring devices should show something like "Can't connect to the server..." in UI when they can't hear from the server for a long time

**Heresy idea:** Maybe don't use device net for suit sensors and replace them just by C# function call? C# function calls should be much cheaper than packet serialize/deserialize +raising events.

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+            if (server == null)
+                return;
+
+            _activeServer = server.Owner;
+            _deviceNetworkSystem.ConnectDevice(server.Owner);
+        }
+
+        if (_activeServer == null)
+            return;
```
Comment (metalgearsloth) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs:64]: Making this a singleton seems bad.

Comment (juliangiebel) [reply Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs:64]: I'll most likely make this one server per station/grid so it wouldn't be a singleton anymore at that point.

**Comment by juliangiebel** [comment]
> I understand this solution for performance reasons, but have some feedback gameplay-wise:
> 
> * When you leave 500 tiles range from a server, your suit sensor won't work. And even if you build another one on your grid it won't help, because all crew monitorings are connected to one world server. So if you have two stations with 500 tiles distance between them - one station won't be able to use their crew monitoring devices.
> 
> * There is no way to build a server. You probably need construction graph and circuit.
> 
> * Mappers should map it asap after this PR merge.
> 
> * Crew monitoring devices should show something like "Can't connect to the server..." in UI when they can't hear from the server for a long time
> 
> 
> **Heresy idea:** Maybe don't use device net for suit sensors and replace them just by C# function call? C# function calls should be much cheaper than packet serialize/deserialize +raising events.

I'll make this one server per station/grid. Having a range makes some sense then but I should propably increase the range.

I forgot to add construction. I will add construction and the "Can't connect to the server.." text after I get back from work today.

**Comment by juliangiebel** [comment]
> packet serialize/deserialize

It's not serialization. You don't even *need* to split your data into seperate fields in the packets payload, you can just slab your data in as is. (Look at how CrewMonitoringServerSystem.cs) does it.
It's just that if it's not a primitive type or it doesn't have some custom logic players won't be able to see the contents of that field.

In the case of suit sensors it's about reusable code. If I write a check for what station the sender and receiver is on you can use it for other devices too instead of having that in the entity system of each of the monitoring devices.

The suit sensor code for creating a device net packet is a bit missleading tbh. Especially with it calling the process serialization/deserialization.

I can even give crew monitoring it's own device net id so incrementing the frequency for each new station would be possible instad of a check if receiver and sender are on the same grid. At that point I can also change the code that creates the suit sensors packet to just put the data in one field as players wil be unnable to see that packet anyway.

**Comment by juliangiebel** [comment]
This is ready for review

File: Content.Server/DeviceNetwork/Components/StationLimitedNetworkComponent.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+{
+    [RegisterComponent]
+    public sealed class StationLimitedNetworkComponent : Component
+    {
+        /// <summary>
+        /// The station id the device is limited to.
+        /// Uses the grid id until moonys station beacon system is implemented
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public GridId? StationId;
```
Comment (metalgearsloth) [inline Content.Server/DeviceNetwork/Components/StationLimitedNetworkComponent.cs:14]: EntityUid

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringServerComponent.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    /// <summary>
+    ///     Whether the server can become the currently active server. The server being unavailable usually means that it isn't powered
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    public bool Available = true;
+
+    /// <summary>
+    /// The id of the station the server is on.
+    /// </summary>
+    public GridId? StationId;
```
Comment (metalgearsloth) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringServerComponent.cs:31]: Shouldn't need to cache this.

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        // check update rate
+        _updateDif += frameTime;
+        if (_updateDif < UpdateRate)
+            return;
+        _updateDif = 0f;
```
Comment (metalgearsloth) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs:39]:
```suggestion
        _updateDif -= UpdateRate;
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Replace GridId with EntityUids.

**Review by moonheart08 — DISMISSED** [review]
StationSystem's in, while the beacons are missing they're not necessary for this PR's functionality. So all GridId needs changed up.

**Review by PaulRitter — DISMISSED** [review]
partial review i apparently never submitted so here you go

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs [hunk 5]
```diff
@@ -50,18 +50,24 @@ public override void Update(float frameTime)
         /// <param name="address">The address of the entity that the packet gets sent to. If null, the message is broadcast to all devices on that frequency (except the sender)</param>
         /// <param name="frequency">The frequency to send on</param>
         /// <param name="data">The data to be sent</param>
-        public void QueuePacket(EntityUid uid, string? address, NetworkPayload data, uint? frequency = null, DeviceNetworkComponent? device = null)
+        /// <returns>Returns true when the packet was successfully enqueued.</returns>
+        public bool QueuePacket(EntityUid uid, string? address, NetworkPayload data, uint? frequency = null, DeviceNetworkComponent? device = null)
```
[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:54]: If frequency is null it get's set to the frequency the device network component has.
If the address is null the package get broadcasted. Both can be null at the same time.

[the remark this answers is not in the document]
Comment (PaulRitter) [reply Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:54]: makes sense

**Comment by juliangiebel** [comment]
I addressed the reviews and made it use station system.
Now There can be one active server per station and the suit sensors get assigned to a station on component init.

**Comment by juliangiebel** [comment]
This is ready to be merged again

**Review by ElectroJr — DISMISSED** [review]
Just some minor guff, otherwise seems good to merge. 

Though for whoever hits the button: they should ping mappers because this will effectively break sensors if maps don't get updated before the next publish.

**Comment by juliangiebel** [comment]
I finally got around to fixing up this PR.

File: Content.Server/DeviceNetwork/Systems/StationLimitedNetworkSystem.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace Content.Server.DeviceNetwork.Systems
+{
+    /// <summary>
+    /// This system requires the StationLimitedNetworkComponent to be on the the sending entity as well as the receiving entity
+    /// </summary>
+    [UsedImplicitly]
+    public sealed class StationLimitedNetworkSystem : EntitySystem
+    {
+        [Dependency] private readonly StationSystem? _stationSystem = default!;
```
Comment (mirrorcult) [inline Content.Server/DeviceNetwork/Systems/StationLimitedNetworkSystem.cs:14]:
```suggestion
        [Dependency] private readonly StationSystem _stationSystem = default!;
```
will always be valid for entity systems

File: Content.Server/DeviceNetwork/Systems/StationLimitedNetworkSystem.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+            component.StationId = stationId;
+        }
+
+        /// <summary>
+        /// Set the station id to the one the entity is on when the station limited component is added
+        /// </summary>
+        private void OnMapInit(EntityUid uid, StationLimitedNetworkComponent networkComponent, MapInitEvent args)
+        {
+            networkComponent.StationId = _stationSystem?.GetOwningStation(uid);
```
Comment (mirrorcult) [inline Content.Server/DeviceNetwork/Systems/StationLimitedNetworkSystem.cs:38]:
```suggestion
            networkComponent.StationId = _stationSystem.GetOwningStation(uid);
```

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs [hunk 8]
```diff
@@ -93,6 +117,11 @@ private void OnMapInit(EntityUid uid, SuitSensorComponent component, MapInitEven
             }
         }
 
+        private void OnComponentInit(EntityUid uid, SuitSensorComponent component, ComponentInit args)
+        {
+            component.StationId = _stationSystem.GetOwningStation(uid);
+        }
```
Comment (mirrorcult) [inline Content.Server/Medical/SuitSensors/SuitSensorSystem.cs:123]: why not on mapinit as well?

File: Resources/Prototypes/Entities/Structures/Machines/crew_monitor_server.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
+﻿- type: entity
+  id: CrewMonitoringServer
+  parent: BaseMachinePowered
+  name: crew monitoring server
+  description: Receives and relais the status of all active suit sensors on the station
```
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Structures/Machines/crew_monitor_server.yml:5]:
```suggestion
  description: Receives and relays the status of all active suit sensors on the station.
```

**Review by mirrorcult — DISMISSED** [review]
this will need to be mapped into the bridge correct?

File: Content.Server/Medical/SuitSensors/SuitSensorSystem.cs (hunk 8 above)
Comment (juliangiebel) [reply Content.Server/Medical/SuitSensors/SuitSensorSystem.cs]: I moved it to mapinit and added a data field annotation

**Comment by juliangiebel** [comment]
@mirrorcult I implemented your suggestions

**Pull request merged**